File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.util.*;
class Solution {
    public int[] solution(String today, String[] terms, String[] privacies) {
        HashMap<String, Integer> map = new HashMap<>();
        ArrayList<Integer> list = new ArrayList<>();
        for(int i = 0; i < terms.length; i++) {
            String[] term = terms[i].split(" ");
            map.put(term[0], Integer.parseInt(term[1]));
        }
        for(int i = 0; i < privacies.length; i++) {
            String[] privacy = privacies[i].split(" ");
            int dueMonth = map.get(privacy[1]);
            int[] date = new int[3];
            String[] sDate = privacy[0].split("\\.");
            for(int j = 0; j < sDate.length; j++) {
                date[j] = Integer.parseInt(sDate[j]);
            }
            int year = date[0];
            int month = date[1];
            int day = date[2];
            year += (month + dueMonth) / 12;
            month = (month + dueMonth) % 12;
            day -= 1;
            if (month == 0) {
                month = 12;
                year -= 1;
            }
            if (day == 0) {
                day = 28;
                month -=1;
            }
            String[] todays = today.split("\\.");
            int tYear = Integer.parseInt(todays[0]);
            int tMonth = Integer.parseInt(todays[1]);
            int tDay = Integer.parseInt(todays[2]);
            if(tYear > year || (tYear == year && tMonth > month) || (tYear == year && tMonth == month && tDay > day)) {
                list.add(i+1);
            }
            
            
            //System.out.print(year + " " + month + " " + day + " ");
            
            
        }
        int[] answer = new int[list.size()];
        for(int i =0; i < list.size(); i++) {
            answer[i] = list.get(i);
        }
        return answer;
    }
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.js
function solution(today, terms, privacies) {
    var answer = [];
    
    const term = {};
    for (const t of terms) {
        const [type, month] = t.split(' ');
        term[type] = Number(month);
    }
    const [year, month, day] = today.split('.').map(Number);
    
    for (let i = 0; i < privacies.length; i++) {
        const p = privacies[i];
        const [date, type] = p.split(' ');
        const [pYear, pMonth, pDay] = date.split('.').map(Number);
        let vYear = pYear + Math.floor((pMonth + term[type]) / 12)
        let vMonth = (pMonth + term[type]) % 12;
        if (vMonth === 0) {
            vYear--;
            vMonth = 12;
        }
        let vDay = pDay - 1;
        if (vDay === 0) {
            vMonth--;
            vDay = 28;
        }
        if (vMonth === 0) {
            vMonth = 12;
            vYear--;
        }
      
        if (new Date(year,month,day).getTime() > new Date(vYear,vMonth,vDay).getTime()) answer.push(i+1);
        
    }
    return answer;
}